backstory/src/test/java/com/google/sps/story/DatamuseRequestClientTest.java

```diff
@@ -0,0 +1,337 @@
+    URLStreamHandlerFactory urlStreamHandlerFactory = mock(URLStreamHandlerFactory.class);
+    URL.setURLStreamHandlerFactory(urlStreamHandlerFactory);
+ 
+    httpUrlStreamHandler = new HttpUrlStreamHandler();
+    BDDMockito.given(urlStreamHandlerFactory.createURLStreamHandler("http")).willReturn(httpUrlStreamHandler);
+  }
+
+  /**
+   * Reset the URLConnections of this stream handler before every method,
+   * so that URLConnections injected for other tests do not mess 
+   * with the current test. Code taken from this tutorial: 
+   * https://claritysoftware.co.uk/mocking-javas-url-with-mockito/.
+   */
+  @Before
+  public void reset() {
+    httpUrlStreamHandler.resetConnections();
+  }
+
+  /** 
+   * Check that IllegalArgumentException thrown for a multiple
+   * word noun.
+   */
+  @Test
+  public void multipleWordNoun() throws Exception {
+    final String MULTIWORD_NOUN = "multiword noun";
+    final int DEFAULT_MAX = 5;
+
+    DatamuseRequestClient client = new DatamuseRequestClient(DEFAULT_URL);
+
+    try {
+      client.fetchRelatedAdjectives(MULTIWORD_NOUN, DEFAULT_MAX);
+    } catch (IllegalArgumentException exception) {
+      return; // test should pass if this exception is thrown
+    } 
+
+    Assert.fail("IllegalArgumentException was not thrown.");
+  }
+
+  /** 
+   * Check that IllegalArgumentException thrown
+   * for a noun with whitespace.
+   */
+  @Test
+  public void nounWithWhitespace() throws Exception {
+    final String WHITESPACE_NOUN = "noun\n";
+    final int DEFAULT_MAX = 5;
+
+    DatamuseRequestClient client = new DatamuseRequestClient(DEFAULT_URL);
+
+    try {
+      client.fetchRelatedAdjectives(WHITESPACE_NOUN, DEFAULT_MAX);
+    } catch (IllegalArgumentException exception) {
+      return; // test should pass if this exception is thrown
+    } 
+
+    Assert.fail("IllegalArgumentException was not thrown.");
+  }
+
+  /** 
+   * Check that APINotAvailableException is thrown 
+   * if there is no URLConnection injected (which is equivalent
+   * to if it can't access the Datamuse API).
+   */
+  @Test 
+  public void apiNotAvailable() throws Exception {
+    final String NOUN = "beach";
+    final int DEFAULT_MAX = 5;
+
+    DatamuseRequestClient client = new DatamuseRequestClient(DEFAULT_URL);
+
+    try {
+      client.fetchRelatedAdjectives(NOUN, DEFAULT_MAX);
+    } catch (APINotAvailableException exception) {
+      return; // test should pass if this exception is thrown
+    }
+
+    Assert.fail("APINotAvailableException was not thrown.");    
+  }
+
+  /** 
+   * Check that fetchRelatedAdjectives() throws a RuntimeException
+   * when the JSON can't be parsed with the correct message signaling
+   * it's an issue with the JSON.
+   */
+  @Test 
+  public void jsonCannotBeParsed() throws Exception {
+    final String NOUN = "beach";
+    final int MAX_ADJECTIVES = 5;
+    final String QUERIES = "rel_jjb=" + NOUN + "&max=" + MAX_ADJECTIVES;
+    final String INPUT = "not JSON";
```

Comment:
When I see "input" I think of an input to the client. IMO this would be easier if named RESPONSE, since this is the response of the HTTP request... to be even clearer, you could name this HTTP_RESPONSE_BODY.

Reply:
That makes sense. I think I'll rename it to RESPONSE.

Reply:
Done